content/blog/gatbsy-typescript/index.md

```diff
@@ -0,0 +1,284 @@
+- [gatsby-plugin-graphql-codegen](https://www.gatsbyjs.org/packages/gatsby-plugin-graphql-codegen/) to automatically generate Typescript types out of the underlying graphql data model that Gatsby uses.
+
+After following the instructions you will have most of the tools, the rest is going to be
+related to 1) configuration and 2) migrating js files.
+
+## Configuration
+
+### Typescript
+
+You will need to create a `tsconfig.json` at the root of your project configuring the
+typescript compiler. This can vary a lot depending on your taste but here's the one we are using:
+
+```json
+{
+  "compilerOptions": {
+    "target": "esnext",
+    "module": "commonjs",
+    "jsx": "preserve",
+    "strict": true,
+    "noImplicitAny": true,
+    "strictNullChecks": false,
+    "strictFunctionTypes": true,
+    "noImplicitThis": true,
+    "alwaysStrict": true,
+    "noUnusedLocals": true,
+    "noUnusedParameters": true,
+    "noImplicitReturns": true,
+    "noFallthroughCasesInSwitch": true,
+    "allowSyntheticDefaultImports": true,
+    "esModuleInterop": true
+  },
+  "include": ["src/**/*"]
+}
+```
+
+I wont cover all the options but I do want to cover `strictNullChecks`.
+I found that enabling this option to be a big pain when convining it with the GraphQL codegen tool.
+Most of the automatically generated types from the queries you make around your codebase
+will have a lot of optionals and `Maybe` making it super verbose and sometimes super
+awkard to handle all these in a strict way.
+
+I usually try to use `strictNullChecks: true` for applications because it makes the typechecking
+to provide even more guarantees about your code, but in this particular application I recommend
+you avoid it altogether.
+
+At this stage you might want to `allowJs: true` to avoid having to migrate _all_ your codebase
+at once. But you should probably migrate some of the core files just to verify the setup.
+
+The process is rather simple, any `.js` file that jas `JSX` inside it should be renamed to `.tsx`,
+if not, it should be rename to `.ts`.
+
+Let the compiler tell you what it expects from those freshly migrated files (see next section): spoiler alert,
+it will ask for props definitions and perhaps some dependencies type definitions.
+
+### `declarations.d.ts`
+
+In most Typescript codebases we have some global types or some untyped modules to declare,
+that is why you most likely will need `src/declaration.d.ts`. Let's see a bair minimum one:
+
+```typescript
+// Make typescript aware of this
+// global Gatbsy variable
+declare var __PATH_PREFIX__: string
+
+// Declare that this modules are untyped
+declare module "typography-theme-github"
+declare module "vfile-message"
+// This makes importing svg images
+// typecheck. You might need
+// to declare other types of files
+// used in your codebase
+declare module "*.svg"
+```
+
+And then you should import this file from your `gatsby-browser.js` which
```

Comment:
"import from" sounds a little confusing to me, imho it should be "import this file *on/in* gatsby-browser.js..." 
mainly because import statements can be "import * **from** whatever" which is the other way around

Reply:
done